File: src/net/minecraft/client/render/entity/PlayerEntityRenderer.cpp
#include "net/minecraft/client/render/entity/PlayerEntityRenderer.hpp"
#include <optional>
#include "net/minecraft/block/Block.hpp"
#include "net/minecraft/client/Minecraft.hpp"
#include "net/minecraft/client/font/TextRenderer.hpp"
#include "net/minecraft/client/render/RenderSystem.hpp"
#include "net/minecraft/client/render/RenderType.hpp"
#include "net/minecraft/client/render/Tessellator.hpp"
#include "net/minecraft/client/render/block/BlockRenderManager.hpp"
#include "net/minecraft/client/render/entity/EntityRenderDispatcher.hpp"
#include "net/minecraft/client/render/item/HeldItemRenderer.hpp"
#include "net/minecraft/entity/player/ClientPlayerEntity.hpp"
#include "net/minecraft/entity/player/PlayerEntity.hpp"
#include "net/minecraft/item/ArmorItem.hpp"
#include "net/minecraft/item/Item.hpp"
#include "net/minecraft/item/ItemStack.hpp"
#include "net/minecraft/util/math/MathHelper.hpp"
namespace net::minecraft::client::render::entity {
namespace {
constexpr const char* kArmorTextureNames[] = {"cloth", "chain", "iron", "diamond", "gold"};
void configureArmShape(model::BipedEntityModel* body,
                       model::BipedEntityModel* armorOuter,
                       model::BipedEntityModel* armorInner,
                       bool slimArms) {
 body->configureArms(slimArms);
 armorOuter->configureArms(slimArms);
 armorInner->configureArms(slimArms);
}
bool resolveSlimArms(const net::minecraft::PlayerEntity& player, EntityRenderDispatcher* dispatcher) {
 if(dispatcher == nullptr || dispatcher->textureManager() == nullptr || player.skinUrl.empty()) {
  return player.slimArms;
 }
 if(const std::optional<bool> downloadedSlim = dispatcher->textureManager()->skinSlimArms(player.skinUrl)) {
  return *downloadedSlim;
 }
 return player.slimArms;
}
} // namespace
PlayerEntityRenderer::PlayerEntityRenderer()
    : LivingEntityRenderer(new model::BipedEntityModel(0.0f), 0.5f),
      bipedModel(static_cast<model::BipedEntityModel*>(model.get())),
      armor1(new model::BipedEntityModel(1.0f)),
      armor2(new model::BipedEntityModel(0.5f)) {
}
void PlayerEntityRenderer::render(
    const net::minecraft::Entity& entity, double x, double y, double z, float yaw, float tickDelta) {
 const auto* player = dynamic_cast<const net::minecraft::PlayerEntity*>(&entity);
 if(player == nullptr) {
  LivingEntityRenderer::render(entity, x, y, z, yaw, tickDelta);
  return;
 }
 const ItemStack* handStack = player->inventory.getSelectedItem();
 configureArmShape(bipedModel, armor1.get(), armor2.get(), resolveSlimArms(*player, dispatcher));
 bipedModel->rightArmPose = handStack != nullptr && !handStack->empty();
 armor2->rightArmPose = bipedModel->rightArmPose;
 armor1->rightArmPose = bipedModel->rightArmPose;
 armor2->sneaking = bipedModel->sneaking = player->isSneaking();
 armor1->sneaking = bipedModel->sneaking;
 double renderY = y - static_cast<double>(player->standingEyeHeight);
 if(player->isSneaking() &&
    dynamic_cast<const ::net::minecraft::entity::player::ClientPlayerEntity*>(player) == nullptr) {
  renderY -= 0.125;
 }
 LivingEntityRenderer::render(entity, x, renderY, z, yaw, tickDelta);
 bipedModel->sneaking = false;
 armor2->sneaking = false;
 armor1->sneaking = false;
 bipedModel->rightArmPose = false;
 armor2->rightArmPose = false;
 armor1->rightArmPose = false;
}
void PlayerEntityRenderer::renderHand() {
 if(bipedModel == nullptr) {
  return;
 }
 if(const auto* player = dynamic_cast<const net::minecraft::PlayerEntity*>(
        dispatcher != nullptr ? dispatcher->cameraEntity() : nullptr)) {
  configureArmShape(bipedModel, armor1.get(), armor2.get(), resolveSlimArms(*player, dispatcher));
 }
 bipedModel->handSwingProgress = 0.0f;
 bipedModel->setAngles(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0625f);
 bipedModel->rightArm.render(0.0625f);
}
bool PlayerEntityRenderer::bindTexture(const net::minecraft::LivingEntity& entity, int layer, float tickDelta) {
 (void)tickDelta;
 const auto* player = dynamic_cast<const net::minecraft::PlayerEntity*>(&entity);
 if(player == nullptr || layer < 0 || layer >= 4) {
  return false;
 }
 const ItemStack helmetStack = player->inventory.armor[static_cast<std::size_t>(3 - layer)];
 if(helmetStack.empty()) {
  return false;
 }
 Item* item = helmetStack.getItem();
 auto* armorItem = dynamic_cast<::net::minecraft::item::ArmorItem*>(item);
 if(armorItem == nullptr) {
  return false;
 }
 const int suffix = layer == 2 ? 2 : 1;
 EntityRenderer::bindTexture(std::string("/armor/") + kArmorTextureNames[armorItem->getTextureIndex()] + "_" +
                             std::to_string(suffix) + ".png");
 model::BipedEntityModel* armorModel = layer == 2 ? armor2.get() : armor1.get();
 armorModel->head.visible = layer == 0;
 armorModel->hat.visible = layer == 0;
 armorModel->body.visible = layer == 1 || layer == 2;
 armorModel->rightArm.visible = layer == 1;
 armorModel->leftArm.visible = layer == 1;
 armorModel->rightLeg.visible = layer == 2 || layer == 3;
 armorModel->leftLeg.visible = layer == 2 || layer == 3;
 setDecorationModel(armorModel);
 return true;
}
void PlayerEntityRenderer::applyScale(const net::minecraft::LivingEntity& entity, float tickDelta) {
 (void)entity;
 (void)tickDelta;
 RenderSystem::scale(0.9375f, 0.9375f, 0.9375f);
}
void PlayerEntityRenderer::applyTranslation(const net::minecraft::LivingEntity& entity, double x, double y, double z) {
 const auto* player = dynamic_cast<const net::minecraft::PlayerEntity*>(&entity);
 if(player != nullptr && player->isAlive() && player->isSleeping()) {
  LivingEntityRenderer::applyTranslation(entity,
                                         x + static_cast<double>(player->sleepOffsetX),
                                         y + static_cast<double>(player->sleepOffsetY),
                                         z + static_cast<double>(player->sleepOffsetZ));
  return;
 }
 LivingEntityRenderer::applyTranslation(entity, x, y, z);
}
void PlayerEntityRenderer::applyHandSwingRotation(const net::minecraft::LivingEntity& entity,
                                                  float headBob,
                                                  float bodyYaw,
                                                  float tickDelta) {
 const auto* player = dynamic_cast<const net::minecraft::PlayerEntity*>(&entity);
 if(player != nullptr && player->isAlive() && player->isSleeping()) {
  RenderSystem::rotate(player->getSleepingRotation(), 0.0f, 1.0f, 0.0f);
  RenderSystem::rotate(getDeathYaw(entity), 0.0f, 0.0f, 1.0f);
  RenderSystem::rotate(270.0f, 0.0f, 1.0f, 0.0f);
  return;
 }
 LivingEntityRenderer::applyHandSwingRotation(entity, headBob, bodyYaw, tickDelta);
}
void PlayerEntityRenderer::renderNameTag(const net::minecraft::LivingEntity& entity, double x, double y, double z) {
 const auto* player = dynamic_cast<const net::minecraft::PlayerEntity*>(&entity);
 if(player == nullptr || dispatcher == nullptr || dispatcher->cameraEntity() == nullptr) {
  return;
 }
 if(!Minecraft::isDisplayGui() || player == dispatcher->cameraEntity()) {
  return;
 }
 constexpr float nameScale = 1.6f;
 const float pixelScale = 0.016666668f * nameScale;
 const float distance = player->getDistance(*dispatcher->cameraEntity());
 const float maxDistance = player->isSneaking() ? 32.0f : 64.0f;
 if(distance >= maxDistance) {
  return;
 }
 if(!player->isSneaking()) {
  if(player->isSleeping()) {
   LivingEntityRenderer::renderNameTag(entity, player->name, x, y - 1.5, z, 64);
  } else {
   LivingEntityRenderer::renderNameTag(entity, player->name, x, y, z, 64);
  }
  return;
 }
 font::TextRenderer* textRenderer = getTextRenderer();
 if(textRenderer == nullptr) {
  return;
 }
 RenderSystem::pushMatrix();
 RenderSystem::translate(static_cast<float>(x), static_cast<float>(y) + 2.3f, static_cast<float>(z));
 RenderSystem::rotate(-dispatcher->yaw_, 0.0f, 1.0f, 0.0f);
 RenderSystem::rotate(dispatcher->pitch_, 1.0f, 0.0f, 0.0f);
 RenderSystem::scale(-pixelScale, -pixelScale, pixelScale);
 render::RenderPassScope passScope(render::RenderType::guiTextured());
 Tessellator& tessellator = Tessellator::INSTANCE;
 {
  RenderSystem::disableTexture();
  tessellator.startQuads();
  const int halfWidth = textRenderer->getWidth(player->name) / 2;
  tessellator.color(0.0f, 0.0f, 0.0f, 0.25f);
  tessellator.vertex(-halfWidth - 1, -1.0, 0.0);
  tessellator.vertex(-halfWidth - 1, 8.0, 0.0);
  tessellator.vertex(halfWidth + 1, 8.0, 0.0);
  tessellator.vertex(halfWidth + 1, -1.0, 0.0);
  tessellator.draw();
 }
 {
  RenderSystem::enableTexture();
  textRenderer->draw(player->name, -textRenderer->getWidth(player->name) / 2, 0, 0x20FFFFFF);
 }
 RenderSystem::color4f(1.0f, 1.0f, 1.0f, 1.0f);
 RenderSystem::popMatrix();
}
void PlayerEntityRenderer::renderMore(const net::minecraft::LivingEntity& entity, float tickDelta) {
 const auto* player = dynamic_cast<const net::minecraft::PlayerEntity*>(&entity);
 if(player == nullptr || bipedModel == nullptr || dispatcher == nullptr) {
  return;
 }
 item::HeldItemRenderer* heldItemRenderer = dispatcher->heldItemRenderer();
 if(heldItemRenderer == nullptr) {
  return;
 }
 const ItemStack helmetStack = player->inventory.armor[3];
 if(!helmetStack.empty() && helmetStack.itemId < 256) {
  RenderSystem::pushMatrix();
  bipedModel->head.transform(0.0625f);
  Block* block = Block::BLOCKS[static_cast<std::size_t>(helmetStack.itemId)];
  if(block != nullptr && block::BlockRenderManager::isSideLit(block->getRenderType())) {
   constexpr float scale = 0.625f;
   RenderSystem::translate(0.0f, -0.25f, 0.0f);
   RenderSystem::rotate(180.0f, 0.0f, 1.0f, 0.0f);
   RenderSystem::scale(scale, -scale, scale);
  }
  heldItemRenderer->renderItem(*player, helmetStack);
  RenderSystem::popMatrix();
 }
  if(player->name == "deadmau5" && !player->skinUrl.empty() && bindDownloadedTexture(player->skinUrl, player->texture)) {
  for(int ear = 0; ear < 2; ++ear) {
   const float headYaw = player->prevYaw + (player->yaw - player->prevYaw) * tickDelta -
                         (player->lastBodyYaw + (player->bodyYaw - player->lastBodyYaw) * tickDelta);
   const float headPitch = player->prevPitch + (player->pitch - player->prevPitch) * tickDelta;
   RenderSystem::pushMatrix();
   RenderSystem::rotate(headYaw, 0.0f, 1.0f, 0.0f);
   RenderSystem::rotate(headPitch, 1.0f, 0.0f, 0.0f);
   RenderSystem::translate(0.375f * static_cast<float>(ear * 2 - 1), 0.0f, 0.0f);
   RenderSystem::translate(0.0f, -0.375f, 0.0f);
   RenderSystem::rotate(-headPitch, 1.0f, 0.0f, 0.0f);
   RenderSystem::rotate(-headYaw, 0.0f, 1.0f, 0.0f);
   constexpr float earScale = 1.3333334f;
   RenderSystem::scale(earScale, earScale, earScale);
   bipedModel->renderEars(0.0625f);
   RenderSystem::popMatrix();
  }
 }
 if(!player->playerCapeUrl.empty() && bindDownloadedTexture(player->playerCapeUrl)) {
  RenderSystem::pushMatrix();
  RenderSystem::translate(0.0f, 0.0f, 0.125f);
  const double capeDx = player->prevCapeX + (player->capeX - player->prevCapeX) * static_cast<double>(tickDelta) -
                        (player->prevX + (player->x - player->prevX) * static_cast<double>(tickDelta));
  const double capeDy = player->prevCapeY + (player->capeY - player->prevCapeY) * static_cast<double>(tickDelta) -
                        (player->prevY + (player->y - player->prevY) * static_cast<double>(tickDelta));
  const double capeDz = player->prevCapeZ + (player->capeZ - player->prevCapeZ) * static_cast<double>(tickDelta) -
                        (player->prevZ + (player->z - player->prevZ) * static_cast<double>(tickDelta));
  const float bodyYaw = player->lastBodyYaw + (player->bodyYaw - player->lastBodyYaw) * tickDelta;
  const double sinYaw = MathHelper::sin(bodyYaw * 3.14159265f / 180.0f);
  const double cosYaw = -MathHelper::cos(bodyYaw * 3.14159265f / 180.0f);
  float capeLift = static_cast<float>(capeDy) * 10.0f;
  if(capeLift < -6.0f) {
   capeLift = -6.0f;
  }
  if(capeLift > 32.0f) {
   capeLift = 32.0f;
  }
  float capeAngle = static_cast<float>(capeDx * sinYaw + capeDz * cosYaw) * 100.0f;
  float capeTwist = static_cast<float>(capeDx * cosYaw - capeDz * sinYaw) * 100.0f;
  if(capeAngle < 0.0f) {
   capeAngle = 0.0f;
  }
  const float stepBob =
      player->prevStepBobbingAmount + (player->stepBobbingAmount - player->prevStepBobbingAmount) * tickDelta;
  capeLift += MathHelper::sin((player->prevHorizontalSpeed +
                               (player->horizontalSpeed - player->prevHorizontalSpeed) * tickDelta) *
                              6.0f) *
              32.0f * stepBob;
  if(player->isSneaking()) {
   capeLift += 25.0f;
  }
  RenderSystem::rotate(6.0f + capeAngle / 2.0f + capeLift, 1.0f, 0.0f, 0.0f);
  RenderSystem::rotate(capeTwist / 2.0f, 0.0f, 0.0f, 1.0f);
  RenderSystem::rotate(-capeTwist / 2.0f, 0.0f, 1.0f, 0.0f);
  RenderSystem::rotate(180.0f, 0.0f, 1.0f, 0.0f);
  bipedModel->renderCape(0.0625f);
  RenderSystem::popMatrix();
 }
 ItemStack handStack =
     player->inventory.getSelectedItem() != nullptr ? *player->inventory.getSelectedItem() : ItemStack{};
 if(player->fishHook != nullptr) {
  handStack = ItemStack{280, 1, 0};
 }
 if(!handStack.empty()) {
  RenderSystem::pushMatrix();
  bipedModel->rightArm.transform(0.0625f);
  RenderSystem::translate(-0.0625f, 0.4375f, 0.0625f);
  if(handStack.itemId < 256) {
   Block* block = Block::BLOCKS[static_cast<std::size_t>(handStack.itemId)];
   if(block != nullptr && block::BlockRenderManager::isSideLit(block->getRenderType())) {
    float scale = 0.5f;
    RenderSystem::translate(0.0f, 0.1875f, -0.3125f);
    RenderSystem::rotate(20.0f, 1.0f, 0.0f, 0.0f);
    RenderSystem::rotate(45.0f, 0.0f, 1.0f, 0.0f);
    scale *= 0.75f;
    RenderSystem::scale(scale, -scale, scale);
   } else if(Item* item = handStack.getItem(); item != nullptr && item->isHandheld()) {
    float scale = 0.625f;
    if(item->isHandheldRod()) {
     RenderSystem::rotate(180.0f, 0.0f, 0.0f, 1.0f);
     RenderSystem::translate(0.0f, -0.125f, 0.0f);
    }
    RenderSystem::translate(0.0f, 0.1875f, 0.0f);
    RenderSystem::scale(scale, -scale, scale);
    RenderSystem::rotate(-100.0f, 1.0f, 0.0f, 0.0f);
    RenderSystem::rotate(45.0f, 0.0f, 1.0f, 0.0f);
   } else {
    const float scale = 0.375f;
    RenderSystem::translate(0.25f, 0.1875f, -0.1875f);
    RenderSystem::scale(scale, scale, scale);
    RenderSystem::rotate(60.0f, 0.0f, 0.0f, 1.0f);
    RenderSystem::rotate(-90.0f, 1.0f, 0.0f, 0.0f);
    RenderSystem::rotate(20.0f, 0.0f, 0.0f, 1.0f);
   }
  } else if(Item* item = handStack.getItem(); item != nullptr && item->isHandheld()) {
   float scale = 0.625f;
   if(item->isHandheldRod()) {
    RenderSystem::rotate(180.0f, 0.0f, 0.0f, 1.0f);
    RenderSystem::translate(0.0f, -0.125f, 0.0f);
   }
   RenderSystem::translate(0.0f, 0.1875f, 0.0f);
   RenderSystem::scale(scale, -scale, scale);
   RenderSystem::rotate(-100.0f, 1.0f, 0.0f, 0.0f);
   RenderSystem::rotate(45.0f, 0.0f, 1.0f, 0.0f);
  } else {
   const float scale = 0.375f;
   RenderSystem::translate(0.25f, 0.1875f, -0.1875f);
   RenderSystem::scale(scale, scale, scale);
   RenderSystem::rotate(60.0f, 0.0f, 0.0f, 1.0f);
   RenderSystem::rotate(-90.0f, 1.0f, 0.0f, 0.0f);
   RenderSystem::rotate(20.0f, 0.0f, 0.0f, 1.0f);
  }
  heldItemRenderer->renderItem(*player, handStack);
  RenderSystem::popMatrix();
 }
}
} // namespace net::minecraft::client::render::entity

File: mcp/src/net/minecraft/client/render/entity/PlayerEntityRenderer.java
/*
 * Decompiled with CFR 0.1.1 (FabricMC 57d88659).
 * 
 * Could not load the following classes:
 *  net.fabricmc.api.EnvType
 *  net.fabricmc.api.Environment
 */
package net.minecraft.client.render.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.block.Block;
import net.minecraft.client.Minecraft;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.render.Tessellator;
import net.minecraft.client.render.block.BlockRenderManager;
import net.minecraft.client.render.entity.LivingEntityRenderer;
import net.minecraft.client.render.entity.model.BipedEntityModel;
import net.minecraft.entity.player.ClientPlayerEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ArmorItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.MathHelper;
import org.lwjgl.opengl.GL11;

@Environment(value=EnvType.CLIENT)
public class PlayerEntityRenderer
extends LivingEntityRenderer {
    private BipedEntityModel bipedModel;
    private BipedEntityModel armor1;
    private BipedEntityModel armor2;
    private static final String[] armorTextureNames = new String[]{"cloth", "chain", "iron", "diamond", "gold"};

    public PlayerEntityRenderer() {
        super(new BipedEntityModel(0.0f), 0.5f);
        this.bipedModel = (BipedEntityModel)this.model;
        this.armor1 = new BipedEntityModel(1.0f);
        this.armor2 = new BipedEntityModel(0.5f);
    }

    protected boolean bindTexture(PlayerEntity arg, int i, float f) {
        Item item;
        ItemStack itemStack = arg.inventory.getArmorStack(3 - i);
        if (itemStack != null && (item = itemStack.getItem()) instanceof ArmorItem) {
            ArmorItem armorItem = (ArmorItem)item;
            this.bindTexture("/armor/" + armorTextureNames[armorItem.textureIndex] + "_" + (i == 2 ? 2 : 1) + ".png");
            BipedEntityModel bipedEntityModel = i == 2 ? this.armor2 : this.armor1;
            bipedEntityModel.head.visible = i == 0;
            bipedEntityModel.hat.visible = i == 0;
            bipedEntityModel.body.visible = i == 1 || i == 2;
            bipedEntityModel.rightArm.visible = i == 1;
            bipedEntityModel.leftArm.visible = i == 1;
            bipedEntityModel.rightLeg.visible = i == 2 || i == 3;
            bipedEntityModel.leftLeg.visible = i == 2 || i == 3;
            this.setDecorationModel(bipedEntityModel);
            return true;
        }
        return false;
    }

    public void render(PlayerEntity arg, double d, double e, double f, float g, float h) {
        ItemStack itemStack = arg.inventory.getSelectedItem();
        this.bipedModel.rightArmPose = itemStack != null;
        this.armor2.rightArmPose = this.bipedModel.rightArmPose;
        this.armor1.rightArmPose = this.bipedModel.rightArmPose;
        this.armor2.sneaking = this.bipedModel.sneaking = arg.isSneaking();
        this.armor1.sneaking = this.bipedModel.sneaking;
        double d2 = e - (double)arg.standingEyeHeight;
        if (arg.isSneaking() && !(arg instanceof ClientPlayerEntity)) {
            d2 -= 0.125;
        }
        super.render(arg, d, d2, f, g, h);
        this.bipedModel.sneaking = false;
        this.armor2.sneaking = false;
        this.armor1.sneaking = false;
        this.bipedModel.rightArmPose = false;
        this.armor2.rightArmPose = false;
        this.armor1.rightArmPose = false;
    }

    protected void renderNameTag(PlayerEntity arg, double d, double e, double f) {
        if (Minecraft.isDisplayGui() && arg != this.dispatcher.cameraEntity) {
            float f2;
            float f3 = 1.6f;
            float f4 = 0.016666668f * f3;
            float f5 = arg.getDistance(this.dispatcher.cameraEntity);
            float f6 = f2 = arg.isSneaking() ? 32.0f : 64.0f;
            if (f5 < f2) {
                String string = arg.name;
                if (!arg.isSneaking()) {
                    if (arg.isSleeping()) {
                        this.renderNameTag(arg, string, d, e - 1.5, f, 64);
                    } else {
                        this.renderNameTag(arg, string, d, e, f, 64);
                    }
                } else {
                    TextRenderer textRenderer = this.getTextRenderer();
                    GL11.glPushMatrix();
                    GL11.glTranslatef((float)d + 0.0f, (float)e + 2.3f, (float)f);
                    GL11.glNormal3f(0.0f, 1.0f, 0.0f);
                    GL11.glRotatef(-this.dispatcher.yaw, 0.0f, 1.0f, 0.0f);
                    GL11.glRotatef(this.dispatcher.pitch, 1.0f, 0.0f, 0.0f);
                    GL11.glScalef(-f4, -f4, f4);
                    GL11.glDisable(2896);
                    GL11.glTranslatef(0.0f, 0.25f / f4, 0.0f);
                    GL11.glDepthMask(false);
                    GL11.glEnable(3042);
                    GL11.glBlendFunc(770, 771);
                    Tessellator tessellator = Tessellator.INSTANCE;
                    GL11.glDisable(3553);
                    tessellator.startQuads();
                    int n = textRenderer.getWidth(string) / 2;
                    tessellator.color(0.0f, 0.0f, 0.0f, 0.25f);
                    tessellator.vertex(-n - 1, -1.0, 0.0);
                    tessellator.vertex(-n - 1, 8.0, 0.0);
                    tessellator.vertex(n + 1, 8.0, 0.0);
                    tessellator.vertex(n + 1, -1.0, 0.0);
                    tessellator.draw();
                    GL11.glEnable(3553);
                    GL11.glDepthMask(true);
                    textRenderer.draw(string, -textRenderer.getWidth(string) / 2, 0, 0x20FFFFFF);
                    GL11.glEnable(2896);
                    GL11.glDisable(3042);
                    GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
                    GL11.glPopMatrix();
                }
            }
        }
    }

    protected void renderMore(PlayerEntity arg, float f) {
        ItemStack itemStack;
        float f2;
        ItemStack itemStack2 = arg.inventory.getArmorStack(3);
        if (itemStack2 != null && itemStack2.getItem().id < 256) {
            GL11.glPushMatrix();
            this.bipedModel.head.transform(0.0625f);
            if (BlockRenderManager.isSideLit(Block.BLOCKS[itemStack2.itemId].getRenderType())) {
                float f3 = 0.625f;
                GL11.glTranslatef(0.0f, -0.25f, 0.0f);
                GL11.glRotatef(180.0f, 0.0f, 1.0f, 0.0f);
                GL11.glScalef(f3, -f3, f3);
            }
            this.dispatcher.heldItemRenderer.renderItem(arg, itemStack2);
            GL11.glPopMatrix();
        }
        if (arg.name.equals("deadmau5") && this.bindDownloadedTexture(arg.skinUrl, null)) {
            for (int i = 0; i < 2; ++i) {
                f2 = arg.prevYaw + (arg.yaw - arg.prevYaw) * f - (arg.lastBodyYaw + (arg.bodyYaw - arg.lastBodyYaw) * f);
                float f4 = arg.prevPitch + (arg.pitch - arg.prevPitch) * f;
                GL11.glPushMatrix();
                GL11.glRotatef(f2, 0.0f, 1.0f, 0.0f);
                GL11.glRotatef(f4, 1.0f, 0.0f, 0.0f);
                GL11.glTranslatef(0.375f * (float)(i * 2 - 1), 0.0f, 0.0f);
                GL11.glTranslatef(0.0f, -0.375f, 0.0f);
                GL11.glRotatef(-f4, 1.0f, 0.0f, 0.0f);
                GL11.glRotatef(-f2, 0.0f, 1.0f, 0.0f);
                float f5 = 1.3333334f;
                GL11.glScalef(f5, f5, f5);
                this.bipedModel.renderEars(0.0625f);
                GL11.glPopMatrix();
            }
        }
        if (this.bindDownloadedTexture(arg.playerCapeUrl, null)) {
            GL11.glPushMatrix();
            GL11.glTranslatef(0.0f, 0.0f, 0.125f);
            double d = arg.prevCapeX + (arg.capeX - arg.prevCapeX) * (double)f - (arg.prevX + (arg.x - arg.prevX) * (double)f);
            double d2 = arg.prevCapeY + (arg.capeY - arg.prevCapeY) * (double)f - (arg.prevY + (arg.y - arg.prevY) * (double)f);
            double d3 = arg.prevCapeZ + (arg.capeZ - arg.prevCapeZ) * (double)f - (arg.prevZ + (arg.z - arg.prevZ) * (double)f);
            float f6 = arg.lastBodyYaw + (arg.bodyYaw - arg.lastBodyYaw) * f;
            double d4 = MathHelper.sin(f6 * (float)Math.PI / 180.0f);
            double d5 = -MathHelper.cos(f6 * (float)Math.PI / 180.0f);
            float f7 = (float)d2 * 10.0f;
            if (f7 < -6.0f) {
                f7 = -6.0f;
            }
            if (f7 > 32.0f) {
                f7 = 32.0f;
            }
            float f8 = (float)(d * d4 + d3 * d5) * 100.0f;
            float f9 = (float)(d * d5 - d3 * d4) * 100.0f;
            if (f8 < 0.0f) {
                f8 = 0.0f;
            }
            float f10 = arg.prevStepBobbingAmount + (arg.stepBobbingAmount - arg.prevStepBobbingAmount) * f;
            f7 += MathHelper.sin((arg.prevHorizontalSpeed + (arg.horizontalSpeed - arg.prevHorizontalSpeed) * f) * 6.0f) * 32.0f * f10;
            if (arg.isSneaking()) {
                f7 += 25.0f;
            }
            GL11.glRotatef(6.0f + f8 / 2.0f + f7, 1.0f, 0.0f, 0.0f);
            GL11.glRotatef(f9 / 2.0f, 0.0f, 0.0f, 1.0f);
            GL11.glRotatef(-f9 / 2.0f, 0.0f, 1.0f, 0.0f);
            GL11.glRotatef(180.0f, 0.0f, 1.0f, 0.0f);
            this.bipedModel.renderCape(0.0625f);
            GL11.glPopMatrix();
        }
        if ((itemStack = arg.inventory.getSelectedItem()) != null) {
            GL11.glPushMatrix();
            this.bipedModel.rightArm.transform(0.0625f);
            GL11.glTranslatef(-0.0625f, 0.4375f, 0.0625f);
            if (arg.fishHook != null) {
                itemStack = new ItemStack(Item.STICK);
            }
            if (itemStack.itemId < 256 && BlockRenderManager.isSideLit(Block.BLOCKS[itemStack.itemId].getRenderType())) {
                f2 = 0.5f;
                GL11.glTranslatef(0.0f, 0.1875f, -0.3125f);
                GL11.glRotatef(20.0f, 1.0f, 0.0f, 0.0f);
                GL11.glRotatef(45.0f, 0.0f, 1.0f, 0.0f);
                GL11.glScalef(f2 *= 0.75f, -f2, f2);
            } else if (Item.ITEMS[itemStack.itemId].isHandheld()) {
                f2 = 0.625f;
                if (Item.ITEMS[itemStack.itemId].isHandheldRod()) {
                    GL11.glRotatef(180.0f, 0.0f, 0.0f, 1.0f);
                    GL11.glTranslatef(0.0f, -0.125f, 0.0f);
                }
                GL11.glTranslatef(0.0f, 0.1875f, 0.0f);
                GL11.glScalef(f2, -f2, f2);
                GL11.glRotatef(-100.0f, 1.0f, 0.0f, 0.0f);
                GL11.glRotatef(45.0f, 0.0f, 1.0f, 0.0f);
            } else {
                f2 = 0.375f;
                GL11.glTranslatef(0.25f, 0.1875f, -0.1875f);
                GL11.glScalef(f2, f2, f2);
                GL11.glRotatef(60.0f, 0.0f, 0.0f, 1.0f);
                GL11.glRotatef(-90.0f, 1.0f, 0.0f, 0.0f);
                GL11.glRotatef(20.0f, 0.0f, 0.0f, 1.0f);
            }
            this.dispatcher.heldItemRenderer.renderItem(arg, itemStack);
            GL11.glPopMatrix();
        }
    }

    protected void applyScale(PlayerEntity arg, float f) {
        float f2 = 0.9375f;
        GL11.glScalef(f2, f2, f2);
    }

    public void renderHand() {
        this.bipedModel.handSwingProgress = 0.0f;
        this.bipedModel.setAngles(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0625f);
        this.bipedModel.rightArm.render(0.0625f);
    }

    protected void applyTranslation(PlayerEntity arg, double d, double e, double f) {
        if (arg.isAlive() && arg.isSleeping()) {
            super.applyTranslation(arg, d + (double)arg.sleepOffsetX, e + (double)arg.sleepOffsetY, f + (double)arg.sleepOffsetZ);
        } else {
            super.applyTranslation(arg, d, e, f);
        }
    }

    protected void applyHandSwingRotation(PlayerEntity arg, float f, float g, float h) {
        if (arg.isAlive() && arg.isSleeping()) {
            GL11.glRotatef(arg.getSleepingRotation(), 0.0f, 1.0f, 0.0f);
            GL11.glRotatef(this.getDeathYaw(arg), 0.0f, 0.0f, 1.0f);
            GL11.glRotatef(270.0f, 0.0f, 1.0f, 0.0f);
        } else {
            super.applyHandSwingRotation(arg, f, g, h);
        }
    }
}

